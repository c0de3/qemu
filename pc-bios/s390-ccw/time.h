#ifndef TIME_H
#define TIME_H

static inline u64 get_clock(void)
{
    u64 r;

    asm volatile("stck %0" : "=Q" (r) : : "cc");
    return r;
}

static inline u64 get_time_ms(void)
{
    /* Bit 51 is incremented each microsecond */
    return (get_clock() >> 12) / 1000;
}

static inline u64 get_time_seconds(void)
{
    return (get_time_ms()) / 1000;
}

static inline void yield(void)
{
    asm volatile ("diag 0,0,0x44"
                  : :
                  : "memory", "cc");
}

static inline void sleep(unsigned int seconds)
{
    ulong target = get_time_seconds() + seconds;

    while (get_time_seconds() < target) {
        yield();
    }
}

#endif
