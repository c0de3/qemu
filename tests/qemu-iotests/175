#!/usr/bin/env python
#
# Test migration corner-cases
#
# Copyright (C) Vladimir Sementsov-Ogievskiy 2017
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.
#

import os
import iotests
import time
from iotests import qemu_img

disk = os.path.join(iotests.test_dir, 'disk')

class TestMigrationCornerCases(iotests.QMPTestCase):
    def setUp(self):
        qemu_img('create', '-f', iotests.imgfmt, disk, '10M')
        self.vm = iotests.VM().add_drive(disk)
        self.vm.launch()

    def tearDown(self):
        self.vm.shutdown()
        os.remove(disk)

    def test_migrate_reset_cont_write(self):
        result = self.vm.qmp('migrate', uri='exec:cat>/dev/null')
        self.assert_qmp(result, 'return', {})
        time.sleep(4)

        result = self.vm.qmp('human-monitor-command',
                             command_line='system_reset')
        self.assert_qmp(result, 'return', '')

        result = self.vm.qmp('cont')
        self.assert_qmp(result, 'return', {})

        result = self.vm.qmp('human-monitor-command',
                             command_line='qemu-io drive0 "write 0 512"')
        self.assert_qmp(result, 'return', '')

    def test_migrate_savevm(self):
        result = self.vm.qmp('migrate', uri='exec:cat>/dev/null')
        self.assert_qmp(result, 'return', {})
        time.sleep(4)

        result = self.vm.qmp('human-monitor-command', command_line='savevm')
        self.assert_qmp(result, 'return', '')

    def test_savevm_set_speed_savevm(self):
        for i in range(10):
            result = self.vm.qmp('human-monitor-command', command_line='savevm')
            self.assert_qmp(result, 'return', '')

            result = self.vm.qmp('migrate_set_speed', value=9223372036853727232)
            self.assert_qmp(result, 'return', {})

if __name__ == '__main__':
    iotests.main()
