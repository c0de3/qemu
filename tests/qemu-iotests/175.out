QA output created by 175
Formatting 'TEST_DIR/t.IMGFMT', fmt=IMGFMT size=16777216 encryption-format=luks luks-key-secret=sec0 luks-iter-time=10

== reading whole image ==
read 16777216/16777216 bytes at offset 0
16 MiB, X ops; XX:XX:XX.X (XXX YYY/sec and XXX ops/sec)

== rewriting whole image ==
wrote 16777216/16777216 bytes at offset 0
16 MiB, X ops; XX:XX:XX.X (XXX YYY/sec and XXX ops/sec)

== verify pattern ==
read 16777216/16777216 bytes at offset 0
16 MiB, X ops; XX:XX:XX.X (XXX YYY/sec and XXX ops/sec)

== verify open failure with wrong password ==
can't open: Invalid password, cannot unlock any keyslot
no file open, try 'help open'
*** done
